platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterUtils.java

```diff
@@ -49,7 +70,117 @@ public static String formatNodeName(final long nodeId) {
      * @return its Hash
      */
     @NonNull
-    public static Hash hash(@NonNull final Roster roster) {
+    public static Hash hashOf(@NonNull final Roster roster) {
         return PBJ_RECORD_HASHER.hash(roster, Roster.PROTOBUF);
     }
+
+    /**
+     * Constructs the initial roster based on the given software version and initial state.
+     * The initial roster is constructed by adopting the candidate roster if a software upgrade is detected.
+     * Otherwise, the initial roster is created from the address book.
+     *
+     * @param version      the software version
+     * @param initialState the initial state
+     * @param addressBook  the address book
+     * @return the initial roster
+     */
+    public static @NonNull Roster constructInitialRoster(
+            @NonNull final SoftwareVersion version,
+            @NonNull final ReservedSignedState initialState,
+            @NonNull final AddressBook addressBook) {
+        final boolean softwareUpgrade = detectSoftwareUpgrade(version, initialState.get());
+        final MerkleRoot merkleRoot = initialState.get().getState();
+        final WritableRosterStore rosterStore = merkleRoot.getWritableRosterStore();
+        final Roster candidateRoster = rosterStore.getCandidateRoster();
+
+        return constructRoster(
+                rosterStore,
+                softwareUpgrade,
+                candidateRoster,
+                addressBook,
+                initialState.get().getRound());
+    }
+
+    /**
+     * Creates a new roster from the bootstrap address book.
+     * @param addressBook the address book
+     * @return a new roster
+     */
+    @NonNull
+    public static Roster createRoster(@NonNull final AddressBook addressBook) {
+        Objects.requireNonNull(addressBook, "The AddressBook must not be null.");
+        final List<RosterEntry> rosterEntries = new ArrayList<>(addressBook.getSize());
+        for (int i = 0; i < addressBook.getSize(); i++) {
```

Comment:
no, it's just preference when I don't need any of the fluent iterator APIs.

Reply:
It's *not* just preference, since with a cursory glance, it appears that the iterator will be more efficient than the hash table lookup you're using for each element. Even if it weren't currently optimized to be better than hash lookup, it could feasibly become so in the future, and we'd want such an optimization to apply everywhere.

If it were just a question of style, the alternative with the iterator is still better, IMO

```
        addressBook.iterator().forEachRemaining(address -> {
            rosterEntries.add(toRosterEntry(address));
        });
```

Reply:
hash table lookup?

Reply:
I was referring to `addressBook.getAddress(nodeId)` for each node ID, which has to look the address up in a map. I was incorrect about the iterator being more efficient, though, since it uses a map lookup under the hood, too.

Feel free to resolve this comment